src/Data/List/Properties.agda

```diff
@@ -1021,13 +1021,13 @@ module _ {R : Rel A p} (R? : B.Decidable R) where
 
   derun-reject : ∀ {x y} xs → R x y → derun R? (x ∷ y ∷ xs) ≡ derun R? (y ∷ xs)
   derun-reject {x} {y} xs Rxy with R? x y
-  ... | yes _    = refl
-  ... | no  ¬Rxy = contradiction Rxy ¬Rxy
+  ... | true  because _      = refl
+  ... | false because [¬Rxy] = contradiction Rxy (invert [¬Rxy])
```

Comment:
Why is this better? Naively the previous version 'looked' better.

Reply:
See #2149 but for a TL;DR justification: I have never liked the asymmetry of having case branches which use constructors which mix different synonyms, such as `true because _` with `no ¬p`... *especially* because in each such case here, and for the bonus of a less eager evaluation semantics (and making sure each branch is 'equally' lazy in that aspect), such things can be replaced by an appeal to `invert`.

Since this PR is part of a suite aimed at removing unnecessary `with`, I might suggest, in the spirit, if not the letter of #2149 that we eventually do away with *all* `with`-based analysis of proofs of `Dec`... but this PR exhibits a half-way house which is more 'uniform' wrt both syntax *and* evaluation...?

Reply:
Sure, inconsistent branches are bad - but here they were doing `yes` / `no` which are consistent. The drawback to the expansion is you have to use `invert` now, and didn't have to before. (I don't believe in cross-function uniformity, that seems like overkill and even counter-productive at times.)

So I'm happy with many of the other changes, it's this particular one which puzzles me. (I'm not saying you're wrong, merely that I'm not understanding your justification for the changes here.) I guess I'm not seeing the eager / lazy part.

I don't want to hold up merging this, it doesn't feel like something worth arguing about, but I would like to understand.

Reply:
> I don't want to hold up merging this, 

Thanks for saying this! But I think this is perhaps a self-inflicted wound which I should attempt to patch by reverting these changes, which don't really belong here. 

> ...it doesn't feel like something worth arguing about, but I would like to understand.

I think that they *ought* to happen at some point (one reason I opened #2149 all those months ago... :-(), but here is not the place. So let's discuss separately, and hopefully be able to make progress downstream.

Reply:
Unless I take your cue about cross-function uniformity (which I had also been going for... that's the second part of the changes, towards eliminating `yes`/`no` altogether) and only revert the proofs about `derun`...? But I still think I' should curb my tendency towards PR mission-creep ...

Reply:
Er... now I'm not sure what to do. I've reverted the changes for `derun`, and looked again at the others, and then *not* reverted those (if only by reason of consistency/uniformity between branches). Not sure whether to go forward or back now!

I think that I lacked the courage of my convictions to try to defend the shift from (more) strict to less in these proofs, and indeed, in *proofs* the issue might be moot, but there's a running theme in the history of `Decidable` about the advantages of being able to work only with the ('cheaper') `Bool`ean `does`, compared with having to recompute the whole ('expensive') `proof`... and AFAICT using the constructors `yes` and `no` *forces* recomputation of the `proof` field (because it asks the matcher to discern which of the constructors `Reflects.ofⁿ` `Reflects.ofʸ` is actually outermost around it, even though those are 'mere' proxies for the corresponding `Bool`...). 

In all the example uses in this module, the 'positive' case is lazy (the `proof` field is thrown away/not bound by `_`), while the 'negative' case passes the `proof` to ... `contradiction`, which as observed elsewhere (eg #2243 / #2199 ), does not/*cannot* use its argument(s). 

Now I can fully believe that I have misconstrued the role of strictness here (and indeed, perhaps a truly lazy evaluator could discern ``Reflects.ofⁿ` `Reflects.ofʸ` without having to do any other evaluation of the proof object underneath, but even that is doing too much traversal, eg via `_×-dec_` etc.?), and perhaps any real distinctions can only be observed with very careful profiling, but having had my eyes opened by #1988 , I'm now ... sensitised to definitions being unnecessarily strict. 

As things stand, neither in the *status quo ante* nor in these revised proofs, is there yet any attempt to replace these uses of `with` by a direct, but nevertheless dependent, elimination on `b because p` as a candidate proof of the `Decidable` property at hand, in which not only would strictness wrt the `p` argument be more clearly discernible, but also that such proofs could be directly by `Bool` elimination of `b`... that's properly the content of #2149 in introducing a 'true' dependent eliminator for `Reflects`/`Decidable` proofs...

... but if you're serious about replacing uses of `with` by direct dependent elimination (eg in terms of introducing explicit helper functions rather than having `with` introduce them behind-the-scenes) then eventually such issue will need to be tackled...

Sorry for such a long essay! I don't know how to distil a TL;DR for you!

Reply:
I'd say that the underlying issue was lack of understanding (of this reviewer) of the subtleties of the evaluation semantics (eager / lazy) involved, and the clear desire for things that are "pure" computations to never force the proof, since it's just going to be discarded anyways.

Part of my issue might be my shallow undertanding of all these syntax patterns and the overall effect they have.

I think the conversation has unearthed non-trivial design choices that ought to be better documented. I think your above essay has done a good job of pushing me towards 'your take' on what design is the better one. 

In (unpublished, sigh) work of a long time ago, we used the words 'verdict' and 'rationale' for what is essentially 'does' and 'proof'. Maybe introducing some wrappers around `Dec` that are known to yield only verdicts (i.e. embody being lazy in the proof) would help?  I'm only introducing these new words as a way to not fall into the rut of using old names with fuzzy semantics & thus muddying the actually interesting discussion.

Reply:
I like the 'verdict'/'rationale' separation of concerns!

As to being 'better documented', I think that `Decidable` and friends (esp. `Reflects`...) *still* represent a work-in-progress (with shout-outs to @laMudri for the work he did in pushing things forward in the direction of such separation), with #2149 and #2199 / #2243 being yet further glosses on that. 

Eg. the factorisation of `Decidable` towards delegation to `Reflects` makes one kind of sense (provided that as far as possible, `Reflects` and its machinery is not/never exposed to the user!), but obscures, on the other hand, a 'direct' API of a form such as
```agda
record Decidable A : Set _ where
  constructor justified
  field
    verdict : Bool
    rationale : if verdict then A else ¬ A
```
(perhaps it was like this, once upon a time?)

As to laziness wrt the `rationale` field, I think that the existing machinery (at least) tries to draw out the design space, not least in the definitions of `_×-dec_` etc. but I think you're right that we still don't quite have the 'right' API... and won't for a while yet!